docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,303 @@
+
+- **Dependency Resolution:** The ``initialize_dependencies`` method in
+  ``Core.py`` receives the bears that are to be run and processes bear
+  dependencies using a consumer-based system so that each dependency bear has
+  only one instance per section and file-dict. It returns a set of dependency
+  bears along with those bears that don't have any dependencies or whose
+  dependencies have been resolved (these are the ones that are scheduled to be
+  run). The bears are resolved by traversing each edge of the dependency graph
+  and instantiating dependency bears accordingly one per section. Before the
+  bears are run we initialize the dependency tracking in the ``__init__`` method
+  of the class ``Session`` which is responsible for running coala sessions. The
+  ``dependency_tracker`` object is used while scheduling and cleaning up bears
+  and also while finishing bear tasks.
+- **Task Scheduling:** The bears that have no dependencies or whose
+  dependencies have been resolved, only their tasks will be scheduled for
+  execution. Before executing any task coala looks it up in the cache.
+  In case of a hit, the existing results that are stored in the cache for the
+  corresponding task arguments are called using ``execute_task_with_cache``
+  method. In case of a miss or if coala is run without a cache the task is
+  executed. The bears without any running tasks are cleaned up from the state of
+  an ongoing run by resolving its dependencies, scheduling dependent bears
+  and removing the bear from the ``running_tasks`` dict. When a task belonging
+  to a particular bear completes, then ``_finish_task`` method is responsible
+  for checking whether the bear execution is complete and the result generated
+  from the task have been returned. If any new tasks are to be scheduled then
+  this method is responsible for doing so after the dependencies have been
+  resolved.
+- **result_callback :** Each of the bears scheduled to run for a session has
+  this callback function which is called when the results are available. It
+  should have the following signature ``def result_callback(result)``.
+- **executors:** These are instances of the ``ProcessPoolExecutor`` class
+  provided by the ``concurrent.futures`` module that is used to execute function
+  calls asynchronously using a pool of processes. Custom ``executor`` instances
+  can be passed to functions as parameters to run the bears. If not provided
+  then a ``ProcessPoolExecutor`` is used using the maximum number of processors
+  available on the machine.
+
+Custom parallelization for bears
+--------------------------------
+A bear inheriting from the class FileBear can parallelize tasks for each file
+given. A bear inheriting from the class DependencyBear can parallelize tasks for
+each dependency result. A bear inheriting from the class ProjectBear does not
+parallelize tasks for each file as it runs on the whole codebase given.
+
+Superior Caching
```

Comment:
I feel that **Superior Caching** and **Task Objects** sections should stay together in the tutorial. IMHO it will provide a more complete narrative that is easier to follow.

Reply:
I can't really remember the exact reason for keeping these two separate but they used to be together and then I think Makman at some point asked me to move task object out of the caching section.

Also I think it would be better to explain task objects before dependency management.

Reply:
Yeah task objects play a quite central role ;)